File: DP/jacks_rental/jacks_rental.cpp
#include <vector>
#include <assert.h>
#include <cmath>
#include <memory>
#include <iostream>
#include <limits>

namespace {

using ValueType = std::vector<std::vector<double> >;
using PolicyType = std::vector<std::vector<std::vector<double> > >; 

std::vector<double> Fac(21, 0.0);
std::vector<double> Request1(21, 0.0);
std::vector<double> Request2(21, 0.0);
std::vector<double> Return1(21, 0.0);
std::vector<double> Return2(21, 0.0);

void CalcFac() {
  Fac[0] = 1;
  Fac[1] = 1;
  for(auto i = 2; i<21; ++i) {
    Fac[i] = i * Fac[i-1];
  }
}

double diff(const ValueType& value1, const ValueType& value2) {
  double ret = 0;
  for(size_t site1 = 0; site1<21; ++site1)
    for(size_t site2 = 0; site2<21; ++site2) {
      ret+=std::abs(value1[site1][site2] - value2[site1][site2]);
    }
  return ret;
}

double diff(const PolicyType& policy1, const PolicyType& policy2) {
  double ret = 0;
  for(size_t site1 = 0; site1<21; ++site1)
    for(size_t site2 = 0; site2<21; ++site2) 
      for(size_t action = 0; action<11; ++action){
      ret+=std::abs(policy1[site1][site2][action] 
        - policy2[site1][site2][action]);
    }
  return ret;
}

double Poisson(double lam, int number) {
  return std::exp(-lam) * (std::pow(lam, number)) / (
    Fac[number]+std::numeric_limits<double>::min());
}

void CalcProbTable() {
  for (int i =0; i<21; ++i) {
    Request1[i] = Poisson(3, i);
    Request2[i] = Poisson(4, i);
    Return1[i] = Poisson(3, i);
    Return2[i] = Poisson(2, i);
  }
}

void ValueEstimation(const ValueType& value, const PolicyType& policy, 
    ValueType* value_new) {
  value_new->resize(21);
  for (auto& v: *value_new) {
    v.resize(21, 0.0);
  }
  for (auto site1 = 0; site1<21; site1++) 
    for (auto site2 = 0; site2<21; site2++) {
      auto max_move = std::min(std::min(site1, 5), 20-site2);
      auto max_receive = std::min(std::min(site2, 5), 20-site1);
      for (auto action=-max_move; action<max_receive+1; ++action) {
        auto prob_policy = policy[site1][site2][action+5];
        auto site1_remain = site1 + action;
        auto site2_remain = site2 - action;
        auto reward = 0.0, expect_future_return = 0.0;
        for(auto site1_rent=0; site1_rent<site1_remain+1; ++site1_rent)
          for(auto site2_rent=0; site2_rent<site2_remain+1; ++site2_rent) {
            for(auto site1_return=0; site1_return<20-(site1_remain-site1_rent)+1; 
              ++site1_return)
              for(auto site2_return=0; site2_return<20-(site2_remain-site2_rent)+1;
                ++site2_return) {
                auto prob = Return1[site1_return] * Return2[site2_return] * 
                         Request1[site1_rent] * Request2[site2_rent];
                expect_future_return = expect_future_return + prob * value[
                    site1_return+site1_remain-site1_rent][site2_return+
                    site2_remain-site2_rent];
              }
            auto prob_rent = Request1[site1_rent] * Request2[site2_rent];
            // std::cout<<prob_rent<<std::endl;
            reward = reward + prob_rent *(site1_rent*10.0 + site2_rent*10.0 
            - 2*std::abs(action));
            //std::cout<<reward<<std::endl;
          }
        value_new->at(site1)[site2] = prob_policy*(reward + expect_future_return*0.9) 
          +value_new->at(site1)[site2];
      }
    }
}

void UpdatePolicy(const ValueType& value, PolicyType* policy_new) {
  policy_new->resize(21);
  for (auto& s2: *policy_new) {
    s2.resize(21);
    for (auto& p: s2) {
      p.resize(11, 0);
    }
  }
  for (auto site1 = 0; site1<21; site1++) 
    for (auto site2 = 0; site2<21; site2++) {
      auto max_move = std::min(std::min(site1, 5), 20-site2);
      auto max_receive = std::min(std::min(site2, 5), 20-site1);
      std::vector<double> action_list(11, 0.0);
      double max_expect = 0;
      size_t max_idx = 0;
      for (auto action=-max_move; action<max_receive+1; ++action) {
        auto site1_remain = site1 + action;
        auto site2_remain = site2 - action;
        auto reward = 0.0, expect_future_return = 0.0;
        for(auto site1_rent=0; site1_rent<site1_remain+1; ++site1_rent)
          for(auto site2_rent=0; site2_rent<site2_remain+1; ++site2_rent) {
            for(auto site1_return=0; site1_return<20-(site1_remain-site1_rent)+1; 
              ++site1_return)
              for(auto site2_return=0; site2_return<20-(site2_remain-site2_rent)+1;
                ++site2_return) {
                auto prob = Return1[site1_return] * Return2[site2_return] * 
                         Request1[site1_rent] * Request2[site2_rent];
                expect_future_return = expect_future_return + prob * value[
                    site1_return+site1_remain-site1_rent][site2_return+
                    site2_remain-site2_rent];
              }
          }
        if(expect_future_return > max_expect) {
          max_expect = expect_future_return;
          max_idx = action+5;
        }
      }
      action_list[max_idx] = 1.0;
      policy_new->at(site1)[site2] = action_list;
    }
}

void VisualizeValue(const ValueType& value) {
  for(auto site1 = 0; site1<21; ++site1) {
    for(auto site2 = 0; site2<21; ++site2) {
      std::cout<<value[site1][site2]<<" ";
    }
    std::cout<<std::endl;
  }
}

void VisualizePolicy(const PolicyType& policy) {
  for(auto site1 = 0; site1<21; ++site1) {
    for(auto site2 = 0; site2<21; ++site2) {
      for(auto action=0; action<11; ++action)
        std::cout<<policy[site1][site2][action]<<" ";
      std::cout<<std::endl;
    }
  }
}


void main() {
  CalcFac();
  CalcProbTable();
  ValueType value(21);
  for (auto& v: value) {
    v.resize(21, 0.0);
  }
  PolicyType policy(21);
  for (auto& s2: policy) {
    s2.resize(21);
    for (auto& p: s2) {
      p.resize(11, 1/11.0);
    }
  }

  ValueType value_new;
  while(1) {
    while(1) {
      ValueEstimation(value, policy, &value_new);
      if(diff(value_new, value)<10) break;
      value = std::move(value_new);
    }
    VisualizeValue(value_new);
    std::cout<<std::endl;
    PolicyType policy_new;
    UpdatePolicy(value, &policy_new);
    if(diff(policy_new, policy)<1) break;
    policy = std::move(policy_new);
    VisualizePolicy(policy);
    std::cout<<std::endl;
  }
}

}  // namespace 

int main(int argc, char** argv) {
  main();
  return 1;
}

File: DP/jacks_rental.py
from __future__ import print_function
from __future__ import division
from __future__ import absolute_import

import numpy as np
import time

kSite1Request = 3
kSite1Return = 3
kSite2Request = 4
kSite2Return = 2
kDiscount = 0.9

Request1 = np.zeros([21])
Request2 = np.zeros([21])
Return1 = np.zeros([21])
Return2 = np.zeros([21])
# rental request:
def GetRentalRequest():
  request1 = np.random.poisson(kSite1Request, 1)
  return1 = np.random.poisson(kSite1Return, 1)
  request2 = np.random.poisson(kSite2Request, 1)
  return2 = np.random.poisson(kSite2Return, 1)
  return request1, return1, request2, return2

def calcProbTable():
  for action in range(21):
    Request1[action] = Poisson(kSite1Request, action)
    Request2[action] = Poisson(kSite2Request, action)
    Return1[action] = Poisson(kSite1Return, action)
    Return2[action] = Poisson(kSite2Return, action)

def Poisson(lam, number):
  return np.exp(-lam)*(lam ** number) / np.math.factorial(number)

def ValueEstimation(value, policy):
  value_new = np.zeros([21, 21])
  for site1 in range(21):
    for site2 in range(21):
      # bellman expection backup 
      # v'(s) = E(R + v(s(t+1)))
      # action = [-5,5]
      max_move = np.min([site1, 5, 20-site2])
      max_receive = np.min([site2, 5, 20-site1])
      possible_action = range(-max_move, max_receive+1)
      for action in possible_action:
        prob_policy = policy[site1, site2, action+5]
        site1_remain = site1 + action
        site2_remain = site2 - action
        reward = 0
        expect_future_return = 0
        for site1_rent in range(site1_remain+1):
          for site2_rent in range(site2_remain+1):
            for site1_return in range(20-(site1_remain - site1_rent)+1):
              for site2_return in range(20-(site2_remain - site2_rent)+1):
                prob = Return1[site1_return] * Return2[site2_return] *\
                   Request1[site1_rent] * Request2[site2_rent]
                expect_future_return = expect_future_return + prob*\
                  value[site1_return+site1_remain-site1_rent, \
                   site2_return+site2_remain-site2_rent]
            prob_rent = Request1[site1_rent] * Request2[site2_rent]
            reward = reward + prob_rent*(site1_rent*10 + site2_rent*10 - \
              2*np.abs(action))
        value_new[site1,site2] = prob_policy*(reward + 
          expect_future_return * kDiscount) + \
          value_new[site1, site2]
  return value_new

def UpdatePolicy(value):
  policy_new = np.zeros([21,21,11])
  for site1 in range(21):
    for site2 in range(21):
      max_move = np.min([site1, 5, 20-site2])
      max_receive = np.min([site2, 5, 20-site1])
      possible_action = range(-max_move, max_receive+1)
      action_list = np.zeros([11])
      for action in possible_action:
        site1_remain = site1 + action
        site2_remain = site2 - action
        expect_future_return = 0
        for site1_rent in range(site1_remain+1):
          for site2_rent in range(site2_remain+1):
            for site1_return in range(20-(site1_remain - site1_rent)+1):
              for site2_return in range(20-(site2_remain - site2_rent)+1):
                prob = Return1[site1_return] * Return2[site2_return]*\
                  Request1[site1_rent] * Request2[site2_rent]
                expect_future_return = expect_future_return + prob*\
                  value[site1_return+site1_remain-site1_rent, \
                   site2_return+site2_remain-site2_rent]
        action_list[action+5] = expect_future_return
      idx = np.argmax(action_list)
      print(idx)
      policy_new[site1, site2, idx] = 1.0
    return policy_new

def main():
  calcProbTable()
  value = np.zeros([21, 21])
  policy = np.ones([21, 21, 11]) * (1 / 11)
  while True:
    while True:
      value_new = ValueEstimation(value, policy)
      if np.sum(np.abs(value_new - value))<100:
        break
      value = value_new
    print("value")
    print(value)
    policy_new = UpdatePolicy(value)
    if np.sum(np.abs(policy_new - policy))<10:
      break
    policy = policy_new
    print("policy")
    print(policy)
    print("====================")

if __name__=='__main__':
  main()  